src/library.js

```diff
@@ -4672,13 +4665,120 @@ LibraryManager.library = {
     } else if (match = /at.*:(\d+):\d+/.exec(frame)) {
       // if we are in js, we can use the js line number as the "return address"
       // this should work for wasm2js and fastcomp
-      return +match[1];
+      // we tag the high bit to distinguish this from wasm addresses
+      return 0x80000000 | +match[1];
     } else {
       // return 0 if we can't find any
       return 0;
     }
   },
 
+  // Returns a representation of a call site of the caller of this function, in a manner
+  // similar to __builtin_return_address. If level is 0, we return the call site of the
+  // caller of this function.
+  emscripten_return_address__deps: ['emscripten_get_callstack_js', 'emscripten_generate_pc'],
+  emscripten_return_address: function(level) {
+    var callstack = _emscripten_get_callstack_js(0).split('\n');
+
+    // skip this function and the caller to get caller's return address
+    return _emscripten_generate_pc(callstack[level + 2]);
+  },
+
+  $UNWIND_CACHE: {},
+
+  // Take a snapshot of the stack so that it can be unwound by emscripten_stack_unwind.
+  emscripten_stack_snapshot__deps: ['emscripten_get_callstack_js', 'emscripten_generate_pc', '$UNWIND_CACHE'],
+  emscripten_stack_snapshot: function () {
+    var callstack = _emscripten_get_callstack_js(0).split('\n');
+    callstack.forEach(function (frame) {
+      var pc = _emscripten_generate_pc(frame);
+      if (pc) {
+        UNWIND_CACHE[pc] = frame;
+      }
+    });
+    // Caches the stack snapshot so that emscripten_stack_unwind() can unwind from this spot.
+    UNWIND_CACHE.last_addr = _emscripten_generate_pc(callstack[2]);
+    UNWIND_CACHE.last_stack = callstack;
+  },
+
+  // Unwinds the stack from a given return address at the top of the stack.
+  // This only works when the stack had a snapshot taken by emscripten_stack_snapshot beforehand.
+  emscripten_stack_unwind__deps: ['$UNWIND_CACHE', 'emscripten_generate_pc'],
```

Comment:
Please write an essay in the comments explaining why this is necessary and how it differs from what native does.